File: 2024/OCT/Make Sum Divisible by P.cpp
int minSubarray(vector<int>& nums, int p) {
        int totalsum=0;
        for(auto i:nums)
        totalsum=(totalsum+i)%p;
        if(totalsum%p==0)
        return 0;

        int sum=0,subarraylength=nums.size(),remainder=totalsum%p;
        unordered_map<int,int>m;
        m[0]=-1;
        for(int i=0 ;i<nums.size();i++)
        {
            sum=(sum+nums[i])%p;
            int prevfind=((sum-remainder)%p+p)%p;
            if(m.find(prevfind)!=m.end())
            subarraylength=min(subarraylength,i-m[prevfind]);

            m[sum]=i;

        }

        return subarraylength==nums.size() ?-1 : subarraylength;
        
    }

File: 2025/NOV/Make Sum Divisible by P.java
class Solution {
    public int minSubarray(int[] nums, int p) {
        
        long totalSum = 0;
        for (int x : nums) totalSum += x;

        long remainder = totalSum % p;
        if (remainder == 0) return 0;

        HashMap<Long, Integer> map = new HashMap<>();
        map.put(0L, -1); 
        long prefix = 0;
        int minLen = nums.length;

        for (int i = 0; i < nums.length; i++) {
            prefix = (prefix + nums[i]) % p;

            long target = (prefix - remainder + p) % p;

            if (map.containsKey(target)) {
                minLen = Math.min(minLen, i - map.get(target));
            }

            map.put(prefix, i); 
        }

        return minLen == nums.length ? -1 : minLen;
    }
}